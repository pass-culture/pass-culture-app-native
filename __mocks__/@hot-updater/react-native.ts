import { ReactNode } from 'react'

const getUpdateSource = jest.fn(() => Promise.resolve({ source: 'source' }))

const HotUpdater = jest.fn(() => (app: ReactNode) => app)

export default Object.assign(HotUpdater, {
  getUpdateSource,
})
