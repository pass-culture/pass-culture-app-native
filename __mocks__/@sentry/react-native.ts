import * as ActualSentryModule from '@sentry/react-native'

const MockedSentryModule = jest.genMockFromModule<typeof ActualSentryModule>('@sentry/react-native')

module.exports = {
  ...MockedSentryModule,
}
