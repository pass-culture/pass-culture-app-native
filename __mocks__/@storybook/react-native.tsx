import React from 'react'; // @storybook
import { View } from 'react-native';

const configure = jest.fn();
const getStorybookUI = jest.fn().mockImplementation(() => () => <View />);

export { configure, getStorybookUI };
