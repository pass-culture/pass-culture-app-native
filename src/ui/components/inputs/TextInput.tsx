import React, { forwardRef } from 'react'
import { TextInput as RNTextInput } from 'react-native'
import styled from 'styled-components/native'
import { v4 as uuidv4 } from 'uuid'

import { useHandleFocus } from 'libs/hooks/useHandleFocus'
import { styledButton } from 'ui/components/buttons/styledButton'
import { FlexInputLabel } from 'ui/components/InputLabel/FlexInputLabel'
import { ContainerWithMaxWidth } from 'ui/components/inputs/ContainerWithMaxWidth'
import { LabelContainer } from 'ui/components/inputs/LabelContainer'
import { RequiredLabel } from 'ui/components/inputs/RequiredLabel'
import { Touchable } from 'ui/components/touchable/Touchable'
import { getSpacing, Spacer, Typo } from 'ui/theme'

import { BaseTextInput } from './BaseTextInput'
import { InputContainer } from './InputContainer'
import { getCustomTextInputProps, getRNTextInputProps, TextInputProps } from './types'

const WithRefTextInput: React.ForwardRefRenderFunction<RNTextInput, TextInputProps> = (
  { ...props },
  forwardedRef
) => {
  const { onFocus, onBlur: onBlurDefault, isFocus } = useHandleFocus()
  const nativeProps = getRNTextInputProps(props)
  const customProps = getCustomTextInputProps(props)
  const textInputID = uuidv4()

  const Icon = customProps.rightButton?.icon
  const StyledIcon =
    Icon &&
    styled(Icon).attrs(({ theme }) => ({
      size: theme.icons.sizes.small,
    }))``

  function onBlur() {
    onBlurDefault()
    if (nativeProps.onBlur) {
      // @ts-expect-error pass event later when needed
      nativeProps.onBlur()
    }
  }

  return (
    <ContainerWithMaxWidth>
      {!!customProps.label && (
        <React.Fragment>
          <FlexInputLabel htmlFor={textInputID}>
            <LabelContainer>
              <Typo.Body>{customProps.label}</Typo.Body>
              <RightLabel
                isRequiredField={customProps.isRequiredField}
                rightLabel={customProps.rightLabel}
              />
            </LabelContainer>
          </FlexInputLabel>
          <Spacer.Column numberOfSpaces={2} />
        </React.Fragment>
      )}
      <InputContainer
        isFocus={isFocus}
        isError={customProps.isError}
        isDisabled={customProps.disabled}
        style={customProps.containerStyle}>
        {customProps.leftComponent ? (
          <React.Fragment>
            {customProps.leftComponent}
            <Spacer.Row numberOfSpaces={2} />
          </React.Fragment>
        ) : null}
        <BaseTextInput
          {...nativeProps}
          nativeID={textInputID}
          ref={forwardedRef}
          onFocus={onFocus}
          onBlur={onBlur}
          accessibilityRequired={customProps.isRequiredField}
          accessibilityDescribedBy={customProps.accessibilityDescribedBy}
        />
        {!!(customProps.rightButton && StyledIcon) && (
          <React.Fragment>
            <Spacer.Row numberOfSpaces={2} />
            <IconTouchableOpacity
              testID={customProps.rightButton.testID}
              accessibilityLabel={customProps.rightButton.accessibilityLabel}
              onPress={customProps.rightButton.onPress}>
              <StyledIcon />
            </IconTouchableOpacity>
          </React.Fragment>
        )}
      </InputContainer>
    </ContainerWithMaxWidth>
  )
}

export const TextInput = forwardRef<RNTextInput, TextInputProps>(WithRefTextInput)

type RightLabelProps = {
  isRequiredField?: boolean
  rightLabel?: string
}
const RightLabel = ({ isRequiredField, rightLabel }: RightLabelProps) => {
  if (isRequiredField) return <RequiredLabel />
  if (rightLabel) return <Typo.CaptionNeutralInfo>{rightLabel}</Typo.CaptionNeutralInfo>
  return null
}

const IconTouchableOpacity = styledButton(Touchable)({
  maxWidth: getSpacing(15),
})
