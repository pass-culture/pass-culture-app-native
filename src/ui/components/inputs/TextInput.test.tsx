import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { TextInput } from './TextInput'

const NormalStateTextInput = <TextInput placeholder="placeholder" onChangeText={doNothingFn} />

describe('<TextInput />', () => {
  it('should render correctly when NOT focused', () => {
    const instance = render(NormalStateTextInput)

    expect(instance).toMatchSnapshot()
  })

  it('should render correctly when focused', () => {
    const notFocusedInstance = render(NormalStateTextInput)
    const focusedInstance = render(NormalStateTextInput)

    const input = focusedInstance.getByPlaceholderText('placeholder')
    fireEvent(input, 'focus')

    expect(notFocusedInstance).toMatchDiffSnapshot(focusedInstance)
  })

  it('should render correctly when in error', () => {
    const notErrorInstance = render(NormalStateTextInput)
    const errorInstance = render(
      <TextInput isError={true} placeholder="placeholder" onChangeText={doNothingFn} />
    )

    expect(notErrorInstance).toMatchDiffSnapshot(errorInstance)
  })
})

function doNothingFn() {
  /* do nothing */
}
