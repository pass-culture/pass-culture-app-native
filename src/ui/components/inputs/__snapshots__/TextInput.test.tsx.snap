// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should render correctly when NOT focused 1`] = `
<View
  isFocus={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "#c7c7cc",
        "borderRadius": 22,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 40,
        "maxWidth": 675,
        "paddingBottom": 4,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 4,
        "shadowColor": "#151515",
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "width": 320,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="placeholder"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
      ]
    }
    underlineColorAndroid="transparent"
  >
    <Text
      color="#626262"
      style={
        Array [
          Object {
            "color": "#626262",
            "fontFamily": "Montserrat-Regular",
            "fontSize": 15,
            "fontWeight": "normal",
            "lineHeight": 20,
          },
        ]
      }
    />
  </TextInput>
</View>
`;

exports[`<TextInput /> should render correctly when focused 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,13 +1,13 @@
  <View
-   isFocus={false}
+   isFocus={true}
    style={
      Array [
        Object {
          \\"alignItems\\": \\"center\\",
          \\"backgroundColor\\": \\"#ffffff\\",
-         \\"borderColor\\": \\"#c7c7cc\\",
+         \\"borderColor\\": \\"#eb0055\\",
          \\"borderRadius\\": 22,
          \\"borderStyle\\": \\"solid\\",
          \\"borderWidth\\": 1,
          \\"flexDirection\\": \\"row\\",
          \\"height\\": 40,"
`;

exports[`<TextInput /> should render correctly when in error 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,13 +1,14 @@
  <View
+   isError={true}
    isFocus={false}
    style={
      Array [
        Object {
          \\"alignItems\\": \\"center\\",
          \\"backgroundColor\\": \\"#ffffff\\",
-         \\"borderColor\\": \\"#c7c7cc\\",
+         \\"borderColor\\": \\"#b60025\\",
          \\"borderRadius\\": 22,
          \\"borderStyle\\": \\"solid\\",
          \\"borderWidth\\": 1,
          \\"flexDirection\\": \\"row\\",
          \\"height\\": 40,"
`;
